Ensure loop device management is accurately updated during upgrade/downgrade

gdemonet opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

**Summary**:

**Acceptance criteria**: 


---

Closes: #2982

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

gdemonet commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -207,6 +207,41 @@ Check pillar before highstate:
+
+  {%- else %}
+    {%- set loop_volume_ids = [] %}
+    {%- for volume in pillar.metalk8s.volumes %}
```
Ah, that can't work, this pillar is computed for a minion in particular... I guess the state should handle it on its own.

gdemonet committed:
salt/volumes: Flatten directory structure

This will simplify the introduction of other states depending on
`.installed`, into which we move the systemd unit template and cleanup
script creation.

See: #2982

gdemonet committed:
salt/volumes: Detach devices before (up|down)grade

Since the management of loop devices changed between 2.6 and 2.7
(introduction of systemd units for this purpose), we want to ensure
there will not be duplicated devices pointing to the same sparse file.
To do this, we introduce a "cleanup" formula, which can operate in two
modes, 'upgrade' and 'downgrade' (controlled via pillar).
For upgrade, we manage this cleanup in the `deploy_node` orchestrate,
during the drain.
For downgrade, we cannot change the `deploy_node` orchestrate, so we
manually drain and cleanup from the `downgrade` orchestrate.

See: #2982

gdemonet committed:
salt/volumes: Provision volumes from highstate

This will ensure all loop devices are managed as expected after an
upgrade.
Note that this means invalid Volumes which were not prepared yet will
break upgrade.

Fixes: #2982

gdemonet committed:
salt/utils: Decode bytes to string in Python 3

The `volume_utils` module interacts with a C library, and some values
are `c_char_p` which translates to bytes in Python 3 (instead of string
in Python 2).
This change breaks the blkid probe, which in turns makes the
`metalk8s_volumes.is_prepared` execution function always return False.
Adding `.decode('ascii')` to the relevant places fixes this issue.

See: bddea0b2b7c7e4c7e0e94745a2d94f2438d49a1d
See: #2982

bert-e commented:
# Conflict




There is a conflict between your branch `bugfix/2982-handle-upgrade-for-loop-device-provisioning` and the
destination branch `development/2.7`.

Please resolve the conflict on **the feature branch** (`bugfix/2982-handle-upgrade-for-loop-device-provisioning`).

```bash
 $ git fetch
 $ git checkout origin/bugfix/2982-handle-upgrade-for-loop-device-provisioning
 $ git merge origin/development/2.7
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:bugfix/2982-handle-upgrade-for-loop-device-provisioning
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

gdemonet committed:
salt: Increase timeout when waiting for containers

When creating `salt-master` and `repositories` static pods, we wait for
some containers to be in "Running" state before continuing. The default
timeout being set to 60s, this step may fail if the underlying platform
is too slow.
We raise this timeout to 180s (3 minutes) to avoid unnecessary failures
(if the container comes up earlier, the state `cri.wait_container` state
will exit as soon as possible).

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -70,7 +72,29 @@ Cordon the node:
     - name: {{ node_name }}
 
-{%- if not pillar.orchestrate.get('skip_draining', False) %}
+{%- if not skip_draining or cleanup_loop_devices %}
+  {%- set run_drain = not skip_draining %}
```
Minor: It's not clear the first time you read this :smile: I would prefer to set `run_drain` at the beginning and use this variable directly in the if
```
{%- set run_drain = not pillar.orchestrate.get('skip_draining', False) %}
[...]
{%- if run_drain or cleanup_loop_devices %}
[...]
```

gdemonet replied:
If it's not clear then I'll change :+1:

(done in 780698a)

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -70,7 +72,29 @@ Cordon the node:
+                             | list
+    %}
+    {%- if volumes_to_clean %}
+      {%- do salt.log.warning(
+        'Forcing drain to clean up loop devices for the following Volumes: '
+        ~ volumes_to_clean | map(attribute='metadata.name') | join(', ')
+      ) %}
+      {%- set run_drain = True %}
```
Do we really want to force ?
Could be better/safer to just fail hard at this point, no ?

gdemonet replied:
Not sure: if we fail hard, it means we can never upgrade a single-node cluster with sparse volumes (unless we remove the volumes before upgrade, which is cumbersome). Since such clusters only exist for dev purposes, I prefer to force the drain (and create a service interruption) rather than prevent the upgrade.

I guess another option could be to require some `metalk8s:upgrade:force_drain` pillar value or something along those lines, but I'm not sure it's worth the effort. WDYT @scality/metalk8s ?

TeddyAndrieux replied:
Hmm yes ok you right make sense.

TeddyAndrieux commented on salt/metalk8s/volumes/cleanup.sls:
```diff
@@ -0,0 +1,56 @@
+    {%- for volume in sparse_volumes %}
+      {%- set volume_name = volume.metadata.name %}
+      {%- set volume_id = volume.metadata.uid %}
+      {%- if mode == 'upgrade' %}
```
Wow ! I do not really like this "mode" stuff, I would prefer to check the node version (which is the desired version) to know if we go to a 2.6.x or 2.7.x

gdemonet replied:
OK let me change that :+1:

(done in 780698a)

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2982-handle-upgrade-for-loop-device-provisioning` into
`development/2.7`
* [w/2.8/bugfix/2982-handle-upgrade-for-loop-device-provisioning](https://github.com/scality/metalk8s/compare/development/2.8...w/2.8/bugfix/2982-handle-upgrade-for-loop-device-provisioning) will be merged into `development/2.8` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #fde9a5a1a8af85f43d1579652a857e7293955e8f on the integration branch
`w/2.8/bugfix/2982-handle-upgrade-for-loop-device-provisioning` is merging a branch which is neither the current
branch `bugfix/2982-handle-upgrade-for-loop-device-provisioning` nor the development branch
`development/2.8`.

It is likely due to a rebase of the branch `bugfix/2982-handle-upgrade-for-loop-device-provisioning` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #e20202eae223556aa6cdf6bbb23546357fbb3431 on the integration branch
`w/2.8/bugfix/2982-handle-upgrade-for-loop-device-provisioning` is merging a branch which is neither the current
branch `bugfix/2982-handle-upgrade-for-loop-device-provisioning` nor the development branch
`development/2.8`.

It is likely due to a rebase of the branch `bugfix/2982-handle-upgrade-for-loop-device-provisioning` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -131,6 +154,28 @@ Wait minion available:
+    - require:
+      - salt: Wait minion available
+      - metalk8s_drain: Drain the node
+    - require_in:
+      - salt: Check pillar before etcd deployment
```
Hmm it means this orchestrate will fail if we are not on an etcd node

gdemonet replied:
Added a check

TeddyAndrieux commented on salt/metalk8s/orchestrate/downgrade/init.sls:
```diff
@@ -73,8 +73,60 @@ Upgrade salt-minion on {{ node }}:
+      {%- do salt.log.warning(
+        'Forcing drain to clean up loop devices for the following Volumes: '
+        ~ volumes_to_clean | map(attribute='metadata.name') | join(', ')
+      ) %}
+    {%- else %}
+      {%- set force_drain = False %}
+    {%- endif %}
+  {%- endif %}
+
+  {%- if force_drain %}
+# We force a node drain manually to handle cleanup of loop devices outside
+# of the `deploy_node` orchestrate, since it comes from `dest_version`.
+# The `deploy_node` orchestrate will take care of uncordoning.
+Cordon node {{ node }}:
+  metalk8s_cordon.node_cordoned:
+    - name: {{ node }}
+    - require:
+      - salt: Upgrade salt-minion on {{ node }}
+
+Drain node {{ node }}:
+  metalk8s_drain.node_drained:
+    - name: {{ node }}
+    - ignore_daemonset: True
+    - delete_local_data: True
+    - force: True
+    - require:
+      - metalk8s_cordon: Cordon node {{ node }}
+
+Cleanup loop devices from node {{ node }}:
+  salt.state:
+    - tgt: {{ node }}
+    - saltenv: {{ saltenv }} {#- Use current version of this formula #}
+    - sls:
+      - metalk8s.volumes.cleanup
+    - require:
+      - metalk8s_drain: Drain node {{ node }}
     - require_in:
       - salt: Deploy node {{ node }}
+
+  {%- endif %}
```
We only need this if dest_version < 2.7.0 maybe worth add an if otherwise, we will do some unnecessary cleanup when downgrading from 2.7.1 to 2.7.0 (even if I know the state will basically do nothing because systemd unit already exists)

gdemonet replied:
Actually, this section is already included within an `if` block (which is used to migrate Salt + Python):

```sls
{%- if salt.pkg.version_cmp(dest_version, '2.7.0') == -1 %}
```

(https://github.com/scality/metalk8s/pull/2984/files/c8c911c4b96741a7987292c0f94caab77e4a6a21#diff-b9a4482d485ff3e2dcd700a24619346950f0db61ef5c07cb0fffe6c9541445c8R62)

TeddyAndrieux replied:
Oh ! You right I didn't see :+1:

TeddyAndrieux commented on salt/metalk8s/volumes/cleanup.sls:
```diff
@@ -0,0 +1,56 @@
+# systemd units, for downgrade.
+{%- set target_version = pillar.metalk8s.nodes[grains.id].version %}
+
+include:
+  - .installed
```
Hmm do we really need to include this "installed" when going to version < 2.7.0 ? Because it's weird to me, we want to remove the systemd unit file but first, we ensure the unit file exists

gdemonet replied:
`.installed` only takes care of "dependencies": some packages for formatting devices, but most importantly, it ensures the cleanup script and the systemd unit **template** are available.

When going from 2.7 to 2.6, we only need the template, to make sure the service can be found (not sure how `service.dead` behaves if this file does not exist).
When going from 2.6 to 2.7, we need the cleanup script, which we call manually.

TeddyAndrieux replied:
> When going from 2.7 to 2.6, we only need the template, to make sure the service can be found (not sure how service.dead behaves if this file does not exist).

```
local:
----------
          ID: toto
    Function: service.dead
      Result: True
     Comment: The named service toto is not available
     Started: 13:39:24.705098
    Duration: 26.128 ms
     Changes:   

Summary for local
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:  26.128 ms
```

I got your point but I think if we have the service running then we already have all the dependencies no ?
Anyway, as you want, it's not really important we can include this "installed" it does not hurt

gdemonet replied:
I still need it when upgrading, so maybe I can only include `if target_version >= 2.7.0`?

TeddyAndrieux replied:
Yes maybe the good approach, I think

gdemonet replied:
Done

TeddyAndrieux commented on salt/metalk8s/volumes/cleanup.sls:
```diff
@@ -0,0 +1,56 @@
+      - test: Ensure Python 3 is available
+      - file: Install clean-up script
+
+      {%- else %} {# target_version < 2.7.0 #}
+
+Disable systemd unit for Volume {{ volume_name }}:
+  service.dead:
+    - name: metalk8s-sparse-volume@{{ volume_id }}
+    - enable: false
+    - require:
+      - test: Ensure Python 3 is available
+      - file: Set up systemd template unit for sparse loop device provisioning
+      - file: Install clean-up script
+
+      {%- endif %}
```
At the very end of the for loop we may want to remove the systemd unit file, so that nobody can manually enable back the service, no ?

gdemonet replied:
Maybe, but I'm not sure why someone would want to enable this service manually unless they know what they're doing. And not sure what the issue would be, except that they may get duplicate loop devices (but then again, they would know, no?).

TeddyAndrieux replied:
Hmm I don't know, 
I'm a user, I have some trouble on my platform, I check for "metalk8s" system services I see this one stopped, I start it :cry: :smile: 
Anyway it will happen only if a user do a downgrade and try to debug itself the cluster after this so ...I agree it's not critical and can be handled later

gdemonet replied:
Since I'm reworking this state, I'll add the cleanup of this unit template.
But I'm not sure how you can find the disabled+stopped services:

```
$ # output after a downgrade from this branch to 2.6.0
$ systemctl list-units -a | grep metalk8s
  srv-scality-metalk8s\x2d2.6.0.mount          loaded    active   mounted   /srv/scality/metalk8s-2.6.0
  srv-scality-metalk8s\x2d2.7.0\x2ddev.mount   loaded    active   mounted   /srv/scality/metalk8s-2.7.0-dev
  var-lib-metalk8s-storage-sparse.mount        loaded    active   mounted   /var/lib/metalk8s/storage/sparse
  system-metalk8s\x2dsparse\x2dvolume.slice    loaded    active   active    system-metalk8s\x2dsparse\x2dvolume.slice
```

TeddyAndrieux replied:
Hmm ok you right, you can have it with "systemctl list-unit-files" but yes ok you do not have the service, anyway would be cleaner to remove the unit file and it does not cost much

gdemonet replied:
Done

TeddyAndrieux reviewed (commented):
LGTM :+1: 
(another PR will be needed to clean up dev/2.8 :) )

TeddyAndrieux reviewed (dismissed):
(Oups + approve) :smile:

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.8/bugfix/2982-handle-upgrade-for-loop-device-provisioning` with contents from `bugfix/2982-handle-upgrade-for-loop-device-provisioning`
and `development/2.8`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.8/bugfix/2982-handle-upgrade-for-loop-device-provisioning origin/development/2.8
 $ git merge origin/bugfix/2982-handle-upgrade-for-loop-device-provisioning
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.8/bugfix/2982-handle-upgrade-for-loop-device-provisioning
```

gdemonet commented:
> (another PR will be needed to clean up dev/2.8 :) )

@TeddyAndrieux I'll take care of the clean-up in the integration branch instead, directly in the merge commit, if that's fine with you.

gdemonet commented:
/approve

bert-e commented:
# History mismatch



Merge commit #f925aa5198beb8dfb91854df4647b9ff0ead2d3c on the integration branch
`w/2.8/bugfix/2982-handle-upgrade-for-loop-device-provisioning` is merging a branch which is neither the current
branch `bugfix/2982-handle-upgrade-for-loop-device-provisioning` nor the development branch
`development/2.8`.

It is likely due to a rebase of the branch `bugfix/2982-handle-upgrade-for-loop-device-provisioning` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/8b8ea4dd0c85389da40ad8d909f3840d18430bf1) did not succeed in branch bugfix/2982-handle-upgrade-for-loop-device-provisioning.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/3d981a4b85a9261f333f0ef47fbfcd36b92edd56) did not succeed in branch w/2.8/bugfix/2982-handle-upgrade-for-loop-device-provisioning.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/8b8ea4dd0c85389da40ad8d909f3840d18430bf1) did not succeed in branch bugfix/2982-handle-upgrade-for-loop-device-provisioning.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #b26180de6bcfee8f7a3cf9b2f6baefff4d402f10 on the integration branch
`w/2.8/bugfix/2982-handle-upgrade-for-loop-device-provisioning` is merging a branch which is neither the current
branch `bugfix/2982-handle-upgrade-for-loop-device-provisioning` nor the development branch
`development/2.8`.

It is likely due to a rebase of the branch `bugfix/2982-handle-upgrade-for-loop-device-provisioning` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.7`

* :heavy_check_mark: `development/2.8`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.7`

* :heavy_check_mark: `development/2.8`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



Please check the status of the associated issue None.

Goodbye gdemonet.
